src/mumble/pluginAPI/PluginComponents.h

```diff
@@ -0,0 +1,232 @@
+
+#include <stdint.h>
+#include <stddef.h>
+
+#define STATUS_OK EC_OK
+
+enum PluginFeature {
+	/// None of the below
+	FEATURE_NONE = 0,
+	/// The plugin provides positional data from a game
+	FEATURE_POSITIONAL = 2 << 0,
+	/// The plugin modifies the input/output audio itself
+	FEATURE_AUDIO = 2 << 1
+};
+
+enum TalkingState {
+	INVALID=-1,
+	PASSIVE=0,
+	TALKING,
+	WHISPERING,
+	SHOUTING
+};
+
+enum TransmissionMode {
+	CONTINOUS,
+	VOICE_ACTIVATION,
+	PUSH_TO_TALK
+};
+
+enum ErrorCode {
+	EC_OK = 0,
+	EC_GENERIC_ERROR
+};
+
+enum PositionalDataErrorCode {
+	/// Positional data has been initialized properly
+	PDEC_OK = 0,
+	/// Positional data is temporarily unavailable (e.g. because the corresponding process isn't running) but might be
+	/// at another point in time.
+	PDEC_ERROR_TEMP,
+	/// Positional data is permanently unavailable (e.g. because the respective memory offsets are outdated).
+	PDEC_ERROR_PERM
+};
+
+struct Version {
+	int32_t major;
+	int32_t minor;
+	int32_t patch;
+};
+
+
+typedef enum TalkingState TalkingState_t;
+typedef enum TransmissionMode TransmissionMode_t;
+typedef struct Version Version_t;
+typedef int32_t MumbleConnection_t;
+typedef uint32_t MumbleUserID_t;
+typedef int32_t MumbleChannelID_t;
+typedef enum ErrorCode error_t;
+
+
+struct MumbleAPI {
+	// -------- Memory management --------
+	
+	/// Frees the given pointer.
+	///
+	/// @param pointer The pointer to free
+	/// @returns The error code. If everything went well, STATUS_OK will be returned.
+	error_t (*freeMemory)(void *pointer);
+
+
+	
+	// -------- Getter functions --------
+
+	/// Gets the connection ID of the server the user is currently active on (the user's audio output is directed at).
+	///
+	/// @param[out] connection A pointer to the memory location the ID should be written to
+	/// @returns The error code. If everything went well, STATUS_OK will be returned. Only then it is valid to access the
+	/// 	value of the provided pointer
+	error_t (*getActiveServerConnection)(MumbleConnection_t *connection);
+
+	/// Fills in the information about the local user.
+	///
+	/// @param connection The ID of the server-connection to use as a context
+	/// @param[out] userID A pointer to the memory the user's ID shall be written to
+	/// @returns The error code. If everything went well, STATUS_OK will be returned. Only then the passed pointer
+	/// 	may be accessed
+	error_t (*getLocalUserID)(MumbleConnection_t connection, MumbleUserID_t *userID);
+
+	/// Fills in the information about the given user's name.
+	///
+	/// @param connection The ID of the server-connection to use as a context
+	/// @param userID The user's ID whose name should be obtained
+	/// @param[out] userName A pointer to where the pointer to the allocated string (C-encoded) should be written to. The
+	/// 	allocated memory has to be freed by a call to freeMemory by the plugin eventually. The memory will only be
+	/// 	allocated if this function returns STATUS_OK.
```

Comment:
Handy not necessarily. But atleast clear.

Reply:
I decided against setting it to `nullptr` as this would require an extra statement before every return statement that might get forgotten in the future leading to inconsistent behaviour.
I also won't make any promises towards the state of `userName` in case of a failure. The doc says you may only access it if the function returns `STATUS_OK` and that's what the user should stick to. If he/she decides to do otherwise, that's undefined behaviour.